library-best-practices/JLBP-3.md

```diff
@@ -0,0 +1,45 @@
+[JLBP-3] Use Semantic versioning
+--------------------------------
+
+Definition of semver: https://semver.org
+
+- For the purpose of compliance with semver, the stable surface of a library
+  serves as its "Public API." Annotations can be used to mark features
+  (classes, methods, etc) as unstable, and thus not part of the Public API,
+  such that semver rules don't have to apply. 
+  - Examples of annotations:
+    - Guava uses `@Beta`
+    - grpc-java uses `@ExperimentalApi`
+    - Google-cloud-java uses `@BetaApi`, `@InternalApi`, and
+      `@InternalExtensionOnly`
+  - Library documentation should let users know how to detect usage of
+    features that use these annotations
+  - Even though semver rules don't apply to unstable features, it is
+    recommended to bump a library's minor version if unstable features have
+    surface breakages.
+- Tools are available to help identify accidental incompatibilities within a
+  major version. Examples:
+  - https://lvc.github.io/japi-compliance-checker/ 
+  - http://www.mojohaus.org/clirr-maven-plugin/
+  - Example compatibility report for grpc-core:
+    https://abi-laboratory.pro/index.php?view=timeline&lang=java&l=grpc-core 
+- Examples of breaking changes that require a new major version:
+  - Upgrading to a non-compatible dependency that is exposed through a
+    library's Public API (for dependencies that follow semver, this happens
+    when a dependency is bumped to a higher major version)
+  - Changing a method signature
+  - Removing a deprecated method
+- Examples of additions that require a new minor version:
+  - Upgrading a dependency to a new minor version (compatible, but new
+    features) that is exposed through a library's Public API
+  - Adding a new class
+  - Adding a new method
+- Special case: Bumping up the minimum required Java version doesn't always
+  require a major version bump. Requirements for skipping the major version
+  bump:
+  - Requirement 1: Only a pure change in the minimum required Java version
+    is allowed. Any surface change that breaks users still requires a major
+    version bump.
+  - Requirement 2: The major version can be kept only when the obsoleted
+    version has minimal usage. If an obsoleted version has medium or high
+    usage, a major version bump may be warranted.
```

Comment:
The rules around a new Java version feel odd to me.

- As I just noted in an internal doc, a Java major-version upgrade is, in most ways that are relevant here, more like a semver library's minor-version bump. As such, it doesn't seem like it should _ever_ trigger a major-version bump in its users. What problem does a major-version bump solve? How would it tie into the forthcoming "\[JLBP-5] Avoid creating conflicts between major versions" and its recommendation to change packages for a major-version upgrade? Would using new Java features ever be worth the trouble caused by a new major version of the library? Maybe the only advice we need to give here is to not upgrade the version of Java until you know that most users can upgrade? Maybe we don't need to say anything about that here yet?

- Requirement 1 is the normal requirement for minor-version upgrades, so it feels strange to call it out again but only here.

Reply:
* If some day there was a version of Java that broke a bunch of stuff, maybe supporting that version might merit a major version bump. So, maybe we should call out the motivation for this rule.
* Yeah, you're right about requirement 1. I think I'll remove it.

Reply:
The new text LGTM, thanks.

More thoughts, which don't need to affect anything you have here: (Sorry that I'm just working my way through this now....)

If someone asked about supporting Java versions, I would probably say:

- In the 99% case, keep supporting the old one until usage is minimal.
- If your library is used from Android, distinguish between the language-feature level that you're targeting and the library level that you're targeting (since tools can make new language features work with old versions of Android but mostly can't help with libraries). (Also, consider [Android market share](https://developer.android.com/about/dashboards/).)
- If you do want to add features for (say) Java 12, then try consider putting them in a separate artifact and package that contains only those features.
- You could also consider doing something like Guava has, with parallel versions. But this has its own downsides.
- Otherwise, you might actually be best served by doing only _minor_ version bumps for the immediate future. This ensures that your consumers, like other libraries, can stick with the current version of your library and still work compatibly with consumers who have moved on to the newer versions that require Java 12.
- Our advice on specific Java versions may change as we see how the new Java release schedule plays out.

Reply:
LGTM on the text of the PR. On @cpovirk's points, I would be very cautious about separate artifacts for Java 12, e.g. o parallel versions. It's hard enough to maintain a single artifact, and split packages are becoming even more troublesome than they are now.

If necessary, one could maintain a single artifact in multiple branches; e.g. JUnit 3 and JUnit 4, or servlet 2.5 and servlet 3. Move the minor versions of each as needed for maintenance but keep the packages and groupId:artifactId the same. Thus we force developers to pick one explicitly and avoid getting both into the classpath. I've seen the opposite and it always seems to be a ^*&&*%^#$# of pain and confusion.

Reply:
Sounds like we should consider a JLBP to cover this issue - I'll add it in the candidate doc.